File: c/test/src/PseudoRandom.c

#include "Common.h"
#include "PseudoRandom.h"

#define multiplier (0x5DEECE66DULL)
#define addend     (0x0BULL)
#define mask       ((1ULL << 48) - 1)
#define MIN(x, y) (((x) < (y)) ? (x) : (y))
static int nextInt_(struct PseudoRandom_Data * data);

void PseudoRandom_Init( struct PseudoRandom_Data * data, int64 seed) {
	int i;
	memset( data, 0, sizeof(struct PseudoRandom_Data));
	data->idx = 0;
    data->seed = (seed ^ multiplier) & mask;
    for (i = 0; i < PseudoRandom_BUFFER_SZ; ) {
        int v = nextInt_(data);
        data->buffer[i++ ] = (uint8)(v >> 24);
        data->buffer[i++ ] = (uint8)(v >> 16);
        data->buffer[i++ ] = (uint8)(v >>  8);
        data->buffer[i++ ] = (uint8)(v);
    }
        
}

void PseudoRandom_NextBytes(struct PseudoRandom_Data * data, uint8* bytes, int offset, int length ) {
    int endOffset = offset+length;
    while( offset < endOffset ){
    	int cpySz = MIN( endOffset - offset, PseudoRandom_BUFFER_SZ - data->idx );
		memcpy( &bytes[offset], &data->buffer[ data->idx ], cpySz );
    	offset += cpySz;
    	data->idx += cpySz;
    	if( data->idx == PseudoRandom_BUFFER_SZ ){
    		data->idx = 0;
    	}
    }
}

void PseudoRandom_NextBytesVerify(struct PseudoRandom_Data * data, uint8* bytes, int offset, int length ) {
	int i;
	for( i = 0; i < length; i++ ){
		uint8 exp = data->buffer[ data->idx % PseudoRandom_BUFFER_SZ ];
		uint8 cur = bytes[i+offset];
		data->idx++;
		if( data->idx == PseudoRandom_BUFFER_SZ ){
    		data->idx = 0;
		}
		if( exp != cur ){
			printf("PseudoRandom: mismatch at %d: exp:0x%x != cur:0x%x", i, exp, cur);
			exit(EXIT_FAILURE);
		}
	}
}
    
uint8 PseudoRandom_NextByte(struct PseudoRandom_Data * data) {
    uint8 res = data->buffer[ data->idx++ % PseudoRandom_BUFFER_SZ ];
    if( data->idx == PseudoRandom_BUFFER_SZ ){
    	data->idx = 0;
    }
    return res;
}

static int nextInt_(struct PseudoRandom_Data * data) {
    data->seed = (data->seed * multiplier + addend) & mask;
    return (int)(data->seed >> 16);
}

File: java/ChabuTest/src/org/chabu/PseudoRandom.java
package org.chabu;

public class PseudoRandom {

    private static final int BUFFER_SZ = 0x2000;
	private long seed;
    private byte[] buffer = new byte[BUFFER_SZ];
    private int idx = 0;

    private static final long multiplier = 0x5_DEEC_E66DL;
    private static final long addend     = 0x0BL;
    private static final long mask       = (1L << 48) - 1;

    public PseudoRandom(long seed) {
        this.seed = (seed ^ multiplier) & mask;
        for (int i = 0; i < buffer.length; ) {
        	int v = nextInt_();
           	buffer[i++ ] = (byte)(v >>> 24);
           	buffer[i++ ] = (byte)(v >>> 16);
           	buffer[i++ ] = (byte)(v >>>  8);
           	buffer[i++ ] = (byte)(v);
        }
        
    }

    public void nextBytes(byte[] bytes, int offset, int length ) {
    	int endOffset = offset+length;
    	while( offset < endOffset ){
    		int cpySz = Math.min( endOffset - offset, BUFFER_SZ - idx );
    		System.arraycopy( buffer, idx, bytes, offset, cpySz );
    		offset += cpySz;
    		idx += cpySz;
    		if( idx == BUFFER_SZ ){
    			idx = 0;
    		}
    	}
    }

	public void nextBytesVerify(byte[] bytes, int offset, int length ) {
		for( int i = 0; i < length; i++ ){
			byte exp = buffer[ idx % buffer.length ];
			byte cur = bytes[i+offset];
			idx++;
    		if( idx == BUFFER_SZ ){
    			idx = 0;
    		}
			if( exp != cur ){
				throw new RuntimeException(String.format("mismatch at %d: exp:0x%02X != cur:0x%02X", i, exp, cur ));
			}
		}
	}
    
    public byte nextByte() {
    	byte res = buffer[ idx ];
		idx++;
    	if( idx == BUFFER_SZ ){
    		idx = 0;
    	}
    	return res;
    }

    private int nextInt_() {
    	this.seed = (this.seed * multiplier + addend) & mask;
    	return (int)(this.seed >>> 16);
    }

}
